### ISSUE TITLE
TP11_EJ10

### ISSUE THREAD

**Issue by Jalbertonisalini** [marcelogarberoglio/PI#258]
Hola Marcelo, me falla en el assert de la linea 29. Me puse a probar y vi que el error estaba en la funcion "most popular" pero no termino de enterder por que. Despues hice otra version de la misnma y anduvo. 

![imagen](https://github.com/marcelogarberoglio/PI/assets/127254155/8a78efb2-e6b6-43da-8758-60213fcedfb7)

Lo que me llama la atencion es que hasta la 29 pasaba todo(inlcuido la que llamaba a mostpopular) pero despues con lo de belen fallaba

![imagen](https://github.com/marcelogarberoglio/PI/assets/127254155/88252aa0-3efd-4e49-a485-3d7bbc6ce225)

Esta es la version que fallaba 

Estoy seguro que el error esta en esa porque cuando use una nueva version de mosPopular anduvo todo correctamente

La version que anda:

![imagen](https://github.com/marcelogarberoglio/PI/assets/127254155/91b85788-3c10-43f5-968a-9eddb5c12127)

**Comment by marcelogarberoglio** [comment]
Falla si el más popular es el último.
Si frenás cuando list->tail es NULL, no comparás la popularidad del úttimo, desacreditando el famoso dicho "Los últimos serán los más populares"o algo así.

**Comment by Jalbertonisalini** [comment]
Y habría entonces manera de que la primer versión ande correctamente? (Con alguna modificacion)

**Comment by marcelogarberoglio** [comment]
No, perdón, me apuré, fallaba porque comparabas siempre con el anterior. Igual confunde un poco lo de cortar por lista->tail != NULL, más claro sería si el ciclo fuera 
while ( list != NULL )

**Comment by Jalbertonisalini** [comment]
Pregunto por el list->tail porque aca me aeguro que si entro tengo aunque sea 1 elemento:

![imagen](https://github.com/marcelogarberoglio/PI/assets/127254155/2b6d4f3f-d48a-417a-a6d1-4256d763fb79)

El tema es que la segunda version que te mande (la que tiene los colores)  anda pero la primera no anda en el caso de belen. De ultima te lo pregunto el lunes al terminar la clase. Total encontre otra alternativa que funciona pero me gustaria entender bien por que la otra no lo hacia.

**Comment by marcelogarberoglio** [comment]
Por lo que comenté en mi segunda respuesta: comparás list->tail->count contra lista->count, por lo que en realidad estás devolviendo el más popular entre el último o el penúltimo.
Pero además hay un error grave que hará que te falle el free, y es cuando hacés
lista->tail = ....
ahí estás rompiendo la lista. Siempre que tengas que recorrer tenés que tener una variable auxiliar.

En la vesión corregida hacés lo correcto: asumés que el primero es el más popular y luego te vas quedando con el mayor.

**Comment by Jalbertonisalini** [comment]
Perdón que te moleste con tantas preguntas pero en la versión corregida por qué no falla en el free si cambio a donde apunta list sin usar un auxiliar?

Te pregunto estas xosas porque algunas cosas de listas las hago con un "salto de fe" y a veces andan y no me termina de cerrar por qué.

**Comment by marcelogarberoglio** [comment]
Sí es una auxiliar, ya que list es el parámetro. Entonces internamente se crea una variable para guadar la copia de la dirección recibida. O sea, no es necesario crear una variable auxiliar, ya el parámetro es una variable local a la función, podés hacer que esa variable local en vez de apuntar al primero apunte al segundo, tercero, etc que no va a afectar a la lista original
Podés cambiar list tranquilamente y no pasa nada. Distinto es si cambiás lo APUNTADO por list, o sea list->tail = .... 
ahí sí estás alterando los datos originales.

**Comment by Jalbertonisalini** [comment]
Ahhhhh muchas gracias. Me estaba confundiendo con eso de las copias porue yo pensaba "la función recibe TList lista" entonces recibe un puntero a nodo pero claro, al llamarla y pasarle first no es que recibe "un puntero" sino directamente la dirección a la que apunta FIRST(copia)
